src/components/ListboxGrid/distribute-resources.ts

```diff
@@ -1,61 +1,232 @@
+    itemCount = Math.min(size.dimensionCount - usedCount, maxPerColumn);
+
+    // TODO:
+    // if (ResponsiveState.isSmallDevice) {
+    // On small device all items is in a single column and overflow scrolled
+    // itemCount = size.dimensionCount - usedCount;
+    // } else if (!zoomEnabled) {
+    if (!zoomEnabled) {
+      // Zoom not enabled (mashup), items evenly distributed and overflow scrolled
+      // distribute the remaining items evenly
+      itemCount = Math.ceil((size.dimensionCount - usedCount) / (maxColumns - columns.length));
+      // to make sure all items in the columns are visible QB-2993
+      itemCount = Math.max(maxPerColumn, itemCount);
+      // but don't add more items then we have left
+      itemCount = Math.min(size.dimensionCount - usedCount, itemCount);
+    }
+
+    columns.push({
+      expand: false,
+      itemCount,
+    });
+
+    // Last columns and all items wont fit, show the there is more items not rendered cue
+    const columnsItemsCount = getColumnItemsCount(columns);
+    if (columns.length >= maxColumns && maxPerColumn > 0 && columnsItemsCount < size.dimensionCount) {
+      columns[columns.length - 1].showAll = true;
+      if (BUTTON_HEIGHT + BUTTON_SPACING + ITEM_SPACING > size.height - getHeightOf(maxPerColumn)) {
+        const lastItem = columns[columns.length - 1];
+        columns[columns.length - 1].itemCount = Math.max((lastItem?.itemCount ?? 0) - 1, 0);
+      }
     }
-  } else if (height >= expandLimit) {
-    if (foldedCount > 0) {
-      folded = rescources.slice(0, foldedCount - 1);
-      expanded = rescources.slice(foldedCount - 1, rescources.length);
-    } else {
-      expanded = [...rescources];
-      folded = [];
+
+    if (getColumnItemsCount(columns) < size.dimensionCount && columns.length < maxColumns) {
+      columns = calculateColumns(size, columns, zoomEnabled);
     }
   }
-  // When moving a resource from rest of columns to this (firstColumn) and height is between fold- and expand-limit, we need to add the missing.
-  if (folded.length + expanded.length !== rescources.length && !expandButton) {
-    const foldedAndExpanded = [...folded, ...expanded];
-    const missingResource = rescources.filter((res) => foldedAndExpanded.every((fe) => fe.id !== res.id));
-    expanded = [...expanded, ...missingResource];
+  return columns;
+};
+
+export const balanceColumns = (size: ISize, columns: IColumn[]) => {
+  let collapsedItems = 0;
+  const expanded = columns.filter((column) => column.expand);
+  const collapsed = columns.filter((column) => column.expand === false);
+
+  // If there is room to expand one item in the last column of collapsed items we expand it.
+  // If not we distribute the items as evenly as possible among the columns with collapsed items.
+  if (
+    collapsed.length > 0
+    && !collapsed[collapsed.length - 1].showAll
+    && haveRoomToExpandOne(size, collapsed[collapsed.length - 1])
+  ) {
```

Comment:
Here also, I would prefer removing/shorten the comments and adding a descriptive variable:
```
const canExpand = collapsed.length > 0 && !collapsed[collapsed.length - 1].showAll && haveRoomToExpandOne(size, collapsed[collapsed.length - 1]);
if (canExpand) { …
```